src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Validators/TimePeriodStringValidators.cs

```diff
@@ -0,0 +1,104 @@
+using System.Text.RegularExpressions;
+using FluentValidation;
+using GovUk.Education.ExploreEducationStatistics.Common.Extensions;
+using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using GovUk.Education.ExploreEducationStatistics.Common.Utils;
+using GovUk.Education.ExploreEducationStatistics.Common.Validators.ErrorDetails;
+
+namespace GovUk.Education.ExploreEducationStatistics.Public.Data.Api.Validators;
+
+public static partial class TimePeriodStringValidators
+{
+    private const string ExpectedFormat = "{period}|{code}";
+
+    public static IRuleBuilderOptionsConditions<T, string> TimePeriodString<T>(this IRuleBuilder<T, string> rule)
+    {
+        return rule.Custom((value, context) =>
+        {
+            if (!HasValidFormat(value))
+            {
+                context.AddFailure(
+                    message: ValidationMessages.TimePeriodFormat,
+                    detail: new FormatErrorDetail(value, ExpectedFormat: ExpectedFormat)
+                );
+                return;
+            }
+
+            var timePeriod = new ParsedTimePeriod(value);
+
+            if (!HasValidRange(timePeriod))
+            {
+                context.AddFailure(
+                    message: ValidationMessages.TimePeriodRange,
+                    detail: new RangeErrorDetail(timePeriod.String)
+                );
+            }
+
+            if (!HasAllowedCode(timePeriod))
+            {
+                context.AddFailure(
+                    message: ValidationMessages.TimePeriodAllowedCode,
+                    detail: new AllowedCodeErrorDetail(timePeriod.String)
+                );
+            }
+        });
+    }
+
+    private static readonly IReadOnlySet<TimeIdentifier> AllowedTimeIdentifiers =
+        TimeIdentifierUtils.DataEnums.ToHashSet();
+
+    [GeneratedRegex(@"^[0-9]{4}(\/[0-9]{4})?\|[A-Z0-9]+$", RegexOptions.Compiled, matchTimeoutMilliseconds: 200)]
+    private static partial Regex FormatRegexGenerated();
+
+    private static readonly Regex FormatRegex = FormatRegexGenerated();
+
+    private static bool HasValidFormat(string value) => FormatRegex.IsMatch(value);
+
+    private static bool HasValidRange(ParsedTimePeriod timePeriod)
+    {
+        // The period is not a range.
+        if (!timePeriod.Period.Contains('/'))
+        {
+            return true;
+        }
+
+        var rangeParts = timePeriod.Period.Split('/');
+        var start = rangeParts[0];
+        var end = rangeParts[1];
+
+        return int.TryParse(start, out var startYear)
+               && int.TryParse(end, out var endYear)
+               && startYear < endYear;
```

Comment:
Was a bit on the fence about this. I ended up letting this be more permissive so that we could potentially allow wider period ranges.

It's also not too obvious if end users will interpret time periods in the way that we expect. They may well try to specify ranges like `2020/2026|AY` and expect to get all the academic years back between 2020 and 2026.

On reflection, this probably introduces more trouble than its worth (right now) though so I've changed it back to the 1 year restriction. We might revisit this in the future depending on user feedback.

Reply:
Yeah I think `2020/2026|AY` could cause more confusion at this point because then you don't know if that means:
- `2020/2021|AY`
- `2021/2022|AY`
- `2022/2023|AY`
- `2023/2024|AY`
- `2024/2025|AY`
- `2025/2026|AY`

Or if it means (there is an extra year at the end):
- `2020/2021|AY`
- `2021/2022|AY`
- `2022/2023|AY`
- `2023/2024|AY`
- `2024/2025|AY`
- `2025/2026|AY`
- `2026/2027|AY`

I personally think restricting it to a 1 year range, like you have, is best for now. Until we definitely need the requirement to extend it :)

Reply:
I like the fact you have restricted it so that ranges can only be used with certain codes too. Nice :)